extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedTracer.java

```diff
@@ -5,54 +5,96 @@
   }
 
-  private ExtendedTracer(Tracer delegate) {
-    this.delegate = delegate;
+  /**
+   * Creates a new {@link ExtendedSpanBuilder} with the given span name.
+   *
+   * @param spanName the name of the span
+   * @return the {@link ExtendedSpanBuilder}
+   */
+  @Override
+  public ExtendedSpanBuilder spanBuilder(String spanName) {
+    return new ExtendedSpanBuilder(delegate.spanBuilder(spanName));
   }
 
-  /** Run the provided {@link Runnable} and wrap with a {@link Span} with the provided name. */
-  public void run(String spanName, Runnable runnable) {
-    Span span = delegate.spanBuilder(spanName).startSpan();
-    try (Scope scope = span.makeCurrent()) {
-      runnable.run();
-    } catch (Throwable e) {
-      span.recordException(e);
-      throw e;
-    } finally {
-      span.end();
-    }
+  /**
+   * Runs the given {@link SpanRunnable} in a new span with the given name and with kind INTERNAL.
+   *
+   * <p>If an exception is thrown by the {@link SpanRunnable}, the span will be marked as error, and
+   * the exception will be recorded.
+   *
+   * @param spanName the name of the span
+   * @param runnable the {@link SpanRunnable} to run
+   * @param <E> the type of the exception
+   */
+  public <E extends Throwable> void run(String spanName, SpanRunnable<E> runnable) throws E {
+    spanBuilder(spanName).run(runnable);
   }
 
-  /** Call the provided {@link Callable} and wrap with a {@link Span} with the provided name. */
-  public <T> T call(String spanName, Callable<T> callable) throws Exception {
-    Span span = delegate.spanBuilder(spanName).startSpan();
-    try (Scope scope = span.makeCurrent()) {
-      return callable.call();
-    } catch (Throwable e) {
-      span.recordException(e);
-      throw e;
-    } finally {
-      span.end();
-    }
+  /**
+   * Runs the given {@link SpanCallable} inside a new span with the given name.
+   *
+   * <p>If an exception is thrown by the {@link SpanCallable}, the span will be marked as error, and
+   * the exception will be recorded.
+   *
+   * @param spanName the name of the span
+   * @param spanCallable the {@link SpanCallable} to call
+   * @param <T> the type of the result
+   * @param <E> the type of the exception
+   * @return the result of the {@link SpanCallable}
+   */
+  public <T, E extends Throwable> T call(String spanName, SpanCallable<T, E> spanCallable)
+      throws E {
+    return spanBuilder(spanName).call(spanCallable);
   }
 
-  @Override
-  public SpanBuilder spanBuilder(String spanName) {
-    return delegate.spanBuilder(spanName);
+  /**
+   * Set baggage items inside the given {@link SpanCallable}.
+   *
+   * @param baggage the baggage items to set
+   * @param spanCallable the {@link SpanCallable} to call
+   * @param <T> the type of the result
+   * @param <E> the type of the exception
+   * @return the result of the {@link SpanCallable}
+   */
+  @SuppressWarnings("NullAway")
+  public static <T, E extends Throwable> T callWithBaggage(
```

Comment:
This method seems out of place. Not sure what to do about it.

Reply:
as discussed, let's scope this out into a separate PR

Reply:
Here it is: https://github.com/open-telemetry/opentelemetry-java/pull/6063